tachyon/math/base/semigroups.h

```diff
@@ -216,6 +324,107 @@ class AdditiveSemigroup {
     }
     return ret;
   }
+
+  // scalar: s
+  // bases: [G₀, G₁, ..., Gₙ₋₁]
+  // return: [sG₀, sG₁, ..., sGₙ₋₁]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> ScalingBases(const F& scalar,
+                                            const std::vector<G>& bases) {
+    size_t size = bases.size();
+    std::vector<ReturnTy> ret(size);
+    size_t num_elems_per_thread = GetNumElementsPerThread(bases);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
+      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
+        ret[j] = bases[j].ScalarMul(scalar.ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalars: [s₀, s₁, ..., sₙ₋₁]
+  // base: G
+  // return: [s₀G, s₁G, ..., sₙ₋₁G]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchScalarsWithBase(
```

Comment:
In my opinion, how about "ScaleEachBases"?
(You can resolve this by yourself)

Reply:
How about ScaleBaseWithScalars? or MulScalarsAndBaseInBatch?

Reply:
When thinking about `BatchInverse()`, `ScalarMul()`, `MSM(MultiScalarMultiplication)`, i think it's good to `BatchScalarMul()` and let compiler differentiate the cases (I discussed with @dongchangYoo )